Firestore/core/test/unit/bundle/bundle_serializer_test.cc

```diff
@@ -26,10 +38,165 @@ namespace bundle {
+using local::LocalSerializer;
+using model::DatabaseId;
+using nanopb::ByteString;
+using nanopb::MakeByteString;
+using nanopb::ProtobufParse;
+using remote::Serializer;
+using std::numeric_limits;
+using testutil::Array;
+using testutil::Filter;
+using testutil::Map;
+using testutil::OrderBy;
+using testutil::Value;
+
+class BundleSerializerTest : public ::testing::Test {
+ public:
+  BundleSerializerTest()
+      : remote_serializer(DatabaseId("p", "default")),
+        local_serializer(remote_serializer),
+        bundle_serializer(remote_serializer) {
+    msg_diff_.ReportDifferencesToString(&message_differences);
+  }
+
+  remote::Serializer remote_serializer;
+  local::LocalSerializer local_serializer;
+  bundle::BundleSerializer bundle_serializer;
+
+  static std::string GetDocumentFullPath(const std::string& path) {
+    return "projects/p/databases/default/documents/" + path;
+  }
+
+  static ProtoDocument TestDocument(ProtoValue value) {
+    ProtoDocument document;
+    document.set_name(GetDocumentFullPath("bundle/test_doc"));
+    auto now = Timestamp::Now();
+    document.mutable_update_time()->set_nanos(now.nanoseconds());
+    document.mutable_update_time()->set_seconds(now.seconds());
+    document.mutable_fields()->insert({"foo", value});
+
+    return document;
+  }
+
+  // 1. Take the value, put it in a libprotobuf Document message and print it
+  // into a JSON string.
+  // 2. Use BundleSerializer to parse the string, then encode the parsed
+  // document into nanopb bytes.
+  // 3. Parse the nanopb bytes to libprotobuf Document message, then compare
+  // with the original.
+  void VerifyFieldValueRoundtrip(ProtoValue value) {
+    ProtoDocument document = TestDocument(value);
+
+    std::string json_string;
+    MessageToJsonString(document, &json_string);
+
+    auto actual = VerifyJsonStringDecodes(json_string);
+
+    VerifyDecodedDocumentEncodesToOriginal(actual.document(), document);
+  }
+
+  void VerifyDecodedDocumentEncodesToOriginal(const model::Document& decoded,
+                                              const ProtoDocument& original) {
+    ByteString bytes =
+        nanopb::MakeByteString(local_serializer.EncodeMaybeDocument(decoded));
+    ProtoMaybeDocument maybe_document;
+    *maybe_document.mutable_document() = original;
+    EXPECT_TRUE(msg_diff_.Compare(maybe_document,
+                                  ProtobufParse<ProtoMaybeDocument>(bytes)))
+        << message_differences;
+    message_differences.clear();
+  }
+
+  void VerifyFieldValueDecodeFails(ProtoValue value) {
+    ProtoDocument document = TestDocument(value);
+
+    std::string json_string;
+    MessageToJsonString(document, &json_string);
+    VerifyJsonStringDecodeFails(std::move(json_string));
+  }
+
+  BundleDocument VerifyJsonStringDecodes(std::string json_string) {
+    JsonReader reader;
+    BundleDocument actual =
+        bundle_serializer.DecodeDocument(reader, json_string);
+    EXPECT_TRUE(reader.ok());
```

Comment:
If this fails you'll get a terrible error message. Instead use `EXPECT_OK` or `EXPECT_NOT_OK` on the underlying status, which will show the error messages if applicable.

Applies throughout.

Reply:
Done.